# 0.0 Notebook 00 - EDA and Cleaning the Raw Data

This is the first step of the pipeline. It reads three months of NYC Yellow Taxi trips (January to March 2024) from HDFS, checks the quality and the native precision of the columns, applies the cleaning rules and writes one clean base back to HDFS. Every storage format and precision level in the next notebooks is built from this base, so all experiments use exactly the same trips.

### 0.1 Imports

In [8]:
from pyspark.sql import SparkSession, functions as F

### 0.2 Data Source and Licence

The trip records are published by the NYC Taxi and Limousine Commission (TLC) as monthly Parquet files, together with the zone lookup table `taxi_zone_lookup.csv`The files were downloaded on 29 September 2026. 

There is no named licence. New York City's Open Data Law allows unrestricted use, as long as whoever redistributes the data names the source and the version and describes the changes; the changes made here are the cleaning rules in Step 5. The data is provided "as is", and there are no driver, vehicle or passenger identifiers, so the dataset has no personal data.

## SETUP & ENVORMENT

The data is read using PySpark from HDFS running in pseudo-distributed mode, which is the processing environment used throughout the project. The Spark session utilizes 4 local cores and 4 GB of memory for the driver, representing the virtual machine's limit.

The session timezone is set to UTC. Pickup and dropoff times are stored as timezone-naive timestamps; if Spark were to use the VM's local timezone (Europe/Dublin), the daylight saving time transition on March 31, 2024, could shift the calculated duration of some trips by one hour.

### 0.3- Initialazing Spark Session and Raw Files

The data is read with PySpark from HDFS running in pseudo-distributed mode. The session uses the 4 local cores and 4 GB for the driver.

The session time zone is set to UTC. Pickup and dropoff are stored as timestamps without a time zone, and with the VM's local zone (Europe/Dublin) the clock change on 31 March 2024 could shift some durations by one hour.

In [5]:
from pyspark.sql import SparkSession, functions as F

# Initialize the spark and give it power to compute 
spark = (SparkSession.builder.master("local[4]").appName("ca1-00-inspect")
         .config("spark.driver.memory", "4g")
         .config("spark.sql.session.timeZone", "UTC")
         .getOrCreate())

# print results and path. 
df = spark.read.parquet("hdfs://localhost:9000/ca1/raw/yellow_tripdata_2024-*.parquet")
df.printSchema()
print("Quantity of lines:", df.count())

26/10/06 16:05:09 WARN FileStreamSink: Assume no metadata directory. Error while looking for metadata directory in the path: hdfs://localhost:9000/ca1/raw/yellow_tripdata_2024-*.parquet.
java.net.ConnectException: Call From BDSP/127.0.1.1 to localhost:9000 failed on connection exception: java.net.ConnectException: Connection refused; For more details see:  http://wiki.apache.org/hadoop/ConnectionRefused
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at org.apache.hadoop.util.dynamic.DynConstructors$Ctor.newInstanceChecked(DynConstructors.jav

Py4JJavaError: An error occurred while calling o50.parquet.
: java.net.ConnectException: Call From BDSP/127.0.1.1 to localhost:9000 failed on connection exception: java.net.ConnectException: Connection refused; For more details see:  http://wiki.apache.org/hadoop/ConnectionRefused
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:500)
	at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:481)
	at org.apache.hadoop.util.dynamic.DynConstructors$Ctor.newInstanceChecked(DynConstructors.java:58)
	at org.apache.hadoop.util.dynamic.DynConstructors$Ctor.newInstance(DynConstructors.java:70)
	at org.apache.hadoop.net.NetUtils.wrapWithMessage(NetUtils.java:1021)
	at org.apache.hadoop.net.NetUtils.wrapException(NetUtils.java:875)
	at org.apache.hadoop.ipc.Client.warpIOException(Client.java:1618)
	at org.apache.hadoop.ipc.Client.getRpcResponse(Client.java:1605)
	at org.apache.hadoop.ipc.Client.call(Client.java:1558)
	at org.apache.hadoop.ipc.Client.call(Client.java:1474)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Invoker.invoke(ProtobufRpcEngine2.java:259)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Invoker.invoke(ProtobufRpcEngine2.java:140)
	at jdk.proxy2/jdk.proxy2.$Proxy46.getListing(Unknown Source)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.lambda$getListing$21(ClientNamenodeProtocolTranslatorPB.java:621)
	at org.apache.hadoop.ipc.internal.ShadedProtobufHelper.ipc(ShadedProtobufHelper.java:160)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.getListing(ClientNamenodeProtocolTranslatorPB.java:621)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:437)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeMethod(RetryInvocationHandler.java:170)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invoke(RetryInvocationHandler.java:162)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeOnce(RetryInvocationHandler.java:100)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:366)
	at jdk.proxy2/jdk.proxy2.$Proxy47.getListing(Unknown Source)
	at org.apache.hadoop.hdfs.DFSClient.listPaths(DFSClient.java:1733)
	at org.apache.hadoop.hdfs.DFSClient.listPaths(DFSClient.java:1717)
	at org.apache.hadoop.hdfs.DistributedFileSystem.listStatusInternal(DistributedFileSystem.java:1122)
	at org.apache.hadoop.hdfs.DistributedFileSystem$24.doCall(DistributedFileSystem.java:1197)
	at org.apache.hadoop.hdfs.DistributedFileSystem$24.doCall(DistributedFileSystem.java:1194)
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
	at org.apache.hadoop.hdfs.DistributedFileSystem.listStatus(DistributedFileSystem.java:1204)
	at org.apache.hadoop.fs.Globber.listStatus(Globber.java:128)
	at org.apache.hadoop.fs.Globber.doGlob(Globber.java:291)
	at org.apache.hadoop.fs.Globber.glob(Globber.java:202)
	at org.apache.hadoop.fs.FileSystem.globStatus(FileSystem.java:2226)
	at org.apache.spark.deploy.SparkHadoopUtil.globPath(SparkHadoopUtil.scala:240)
	at org.apache.spark.sql.execution.datasources.DataSource$.$anonfun$checkAndGlobPathIfNecessary$3(DataSource.scala:816)
	at ... run in separate thread using org.apache.spark.util.ThreadUtils ... ()
	at org.apache.spark.sql.execution.datasources.DataSource$.checkAndGlobPathIfNecessary(DataSource.scala:828)
	at org.apache.spark.sql.execution.datasources.DataSource.checkAndGlobPathIfNecessary(DataSource.scala:593)
	at org.apache.spark.sql.execution.datasources.DataSource.resolveRelation(DataSource.scala:436)
	at org.apache.spark.sql.catalyst.analysis.ResolveDataSource.org$apache$spark$sql$catalyst$analysis$ResolveDataSource$$loadV1BatchSource(ResolveDataSource.scala:210)
	at org.apache.spark.sql.catalyst.analysis.ResolveDataSource$$anonfun$apply$1.$anonfun$applyOrElse$2(ResolveDataSource.scala:62)
	at scala.Option.getOrElse(Option.scala:201)
	at org.apache.spark.sql.catalyst.analysis.ResolveDataSource$$anonfun$apply$1.applyOrElse(ResolveDataSource.scala:62)
	at org.apache.spark.sql.catalyst.analysis.ResolveDataSource$$anonfun$apply$1.applyOrElse(ResolveDataSource.scala:46)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.$anonfun$resolveOperatorsUpWithPruning$3(AnalysisHelper.scala:139)
	at org.apache.spark.sql.catalyst.trees.CurrentOrigin$.withOrigin(origin.scala:107)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.$anonfun$resolveOperatorsUpWithPruning$1(AnalysisHelper.scala:139)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.allowInvokingTransformsInAnalyzer(AnalysisHelper.scala:416)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperatorsUpWithPruning(AnalysisHelper.scala:135)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperatorsUpWithPruning$(AnalysisHelper.scala:131)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveOperatorsUpWithPruning(LogicalPlan.scala:37)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperatorsUp(AnalysisHelper.scala:112)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper.resolveOperatorsUp$(AnalysisHelper.scala:111)
	at org.apache.spark.sql.catalyst.plans.logical.LogicalPlan.resolveOperatorsUp(LogicalPlan.scala:37)
	at org.apache.spark.sql.catalyst.analysis.ResolveDataSource.apply(ResolveDataSource.scala:46)
	at org.apache.spark.sql.catalyst.analysis.ResolveDataSource.apply(ResolveDataSource.scala:44)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$2(RuleExecutor.scala:248)
	at scala.collection.LinearSeqOps.foldLeft(LinearSeq.scala:183)
	at scala.collection.LinearSeqOps.foldLeft$(LinearSeq.scala:179)
	at scala.collection.immutable.List.foldLeft(List.scala:79)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$1(RuleExecutor.scala:245)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$execute$1$adapted(RuleExecutor.scala:237)
	at scala.collection.immutable.List.foreach(List.scala:323)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.execute(RuleExecutor.scala:237)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.super$execute(Analyzer.scala:438)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.$anonfun$executeSameContext$1(Analyzer.scala:438)
	at org.apache.spark.sql.internal.SQLConf$.withExistingConf(SQLConf.scala:171)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.runWithSessionConf(Analyzer.scala:400)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.org$apache$spark$sql$catalyst$analysis$Analyzer$$executeSameContext(Analyzer.scala:438)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.$anonfun$execute$1(Analyzer.scala:433)
	at org.apache.spark.sql.catalyst.analysis.AnalysisContext$.withNewAnalysisContext(Analyzer.scala:276)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.execute(Analyzer.scala:433)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.execute(Analyzer.scala:336)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.$anonfun$executeAndTrack$1(RuleExecutor.scala:207)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker$.withTracker(QueryPlanningTracker.scala:89)
	at org.apache.spark.sql.catalyst.rules.RuleExecutor.executeAndTrack(RuleExecutor.scala:207)
	at org.apache.spark.sql.catalyst.analysis.resolver.HybridAnalyzer.resolveInFixedPoint(HybridAnalyzer.scala:273)
	at org.apache.spark.sql.catalyst.analysis.resolver.HybridAnalyzer.$anonfun$apply$1(HybridAnalyzer.scala:82)
	at org.apache.spark.sql.catalyst.analysis.resolver.HybridAnalyzer.withTrackedAnalyzerBridgeState(HybridAnalyzer.scala:117)
	at org.apache.spark.sql.catalyst.analysis.resolver.HybridAnalyzer.apply(HybridAnalyzer.scala:75)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.runAnalysis$1(Analyzer.scala:368)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.$anonfun$executeAndCheck$2(Analyzer.scala:373)
	at org.apache.spark.sql.internal.SQLConf$.withExistingConf(SQLConf.scala:171)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.runWithSessionConf(Analyzer.scala:400)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.$anonfun$executeAndCheck$1(Analyzer.scala:373)
	at org.apache.spark.sql.catalyst.plans.logical.AnalysisHelper$.markInAnalyzer(AnalysisHelper.scala:423)
	at org.apache.spark.sql.catalyst.analysis.Analyzer.executeAndCheck(Analyzer.scala:373)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$lazyAnalyzed$2(QueryExecution.scala:200)
	at org.apache.spark.sql.catalyst.QueryPlanningTracker.measurePhase(QueryPlanningTracker.scala:148)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$3(QueryExecution.scala:410)
	at org.apache.spark.sql.execution.QueryExecution.withQueryExecutionId(QueryExecution.scala:429)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$2(QueryExecution.scala:410)
	at org.apache.spark.sql.execution.QueryExecution$.withInternalError(QueryExecution.scala:872)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$executePhase$1(QueryExecution.scala:409)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:810)
	at org.apache.spark.sql.execution.QueryExecution.executePhase(QueryExecution.scala:408)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$lazyAnalyzed$1(QueryExecution.scala:200)
	at scala.util.Try$.apply(Try.scala:217)
	at org.apache.spark.util.Utils$.doTryWithCallerStacktrace(Utils.scala:1407)
	at org.apache.spark.util.Utils$.getTryWithCallerStacktrace(Utils.scala:1457)
	at org.apache.spark.util.LazyTry.get(LazyTry.scala:61)
	at org.apache.spark.sql.execution.QueryExecution.$anonfun$analyzed$1(QueryExecution.scala:212)
	at org.apache.spark.sql.execution.QueryExecution.withAbortTransactionOnFailure(QueryExecution.scala:632)
	at org.apache.spark.sql.execution.QueryExecution.analyzed(QueryExecution.scala:212)
	at org.apache.spark.sql.execution.QueryExecution.assertAnalyzed(QueryExecution.scala:151)
	at org.apache.spark.sql.classic.Dataset$.$anonfun$ofRows$1(Dataset.scala:114)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:810)
	at org.apache.spark.sql.classic.Dataset$.ofRows(Dataset.scala:112)
	at org.apache.spark.sql.classic.DataFrameReader.load(DataFrameReader.scala:109)
	at org.apache.spark.sql.classic.DataFrameReader.load(DataFrameReader.scala:58)
	at org.apache.spark.sql.DataFrameReader.parquet(DataFrameReader.scala:457)
	at org.apache.spark.sql.classic.DataFrameReader.parquet(DataFrameReader.scala:306)
	at org.apache.spark.sql.classic.DataFrameReader.parquet(DataFrameReader.scala:58)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:374)
	at py4j.Gateway.invoke(Gateway.java:282)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.ClientServerConnection.waitForCommands(ClientServerConnection.java:184)
	at py4j.ClientServerConnection.run(ClientServerConnection.java:108)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: java.net.ConnectException: Connection refused
	at java.base/sun.nio.ch.Net.pollConnect(Native Method)
	at java.base/sun.nio.ch.Net.pollConnectNow(Net.java:684)
	at java.base/sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:946)
	at org.apache.hadoop.net.SocketIOWithTimeout.connect(SocketIOWithTimeout.java:205)
	at org.apache.hadoop.net.NetUtils.connect(NetUtils.java:613)
	at org.apache.hadoop.ipc.Client$Connection.setupConnection(Client.java:715)
	at org.apache.hadoop.ipc.Client$Connection.setupIOstreams(Client.java:836)
	at org.apache.hadoop.ipc.Client.getConnection(Client.java:1675)
	at org.apache.hadoop.ipc.Client.call(Client.java:1520)
	at org.apache.hadoop.ipc.Client.call(Client.java:1474)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Invoker.invoke(ProtobufRpcEngine2.java:259)
	at org.apache.hadoop.ipc.ProtobufRpcEngine2$Invoker.invoke(ProtobufRpcEngine2.java:140)
	at jdk.proxy2/jdk.proxy2.$Proxy46.getListing(Unknown Source)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.lambda$getListing$21(ClientNamenodeProtocolTranslatorPB.java:621)
	at org.apache.hadoop.ipc.internal.ShadedProtobufHelper.ipc(ShadedProtobufHelper.java:160)
	at org.apache.hadoop.hdfs.protocolPB.ClientNamenodeProtocolTranslatorPB.getListing(ClientNamenodeProtocolTranslatorPB.java:621)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invokeMethod(RetryInvocationHandler.java:437)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeMethod(RetryInvocationHandler.java:170)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invoke(RetryInvocationHandler.java:162)
	at org.apache.hadoop.io.retry.RetryInvocationHandler$Call.invokeOnce(RetryInvocationHandler.java:100)
	at org.apache.hadoop.io.retry.RetryInvocationHandler.invoke(RetryInvocationHandler.java:366)
	at jdk.proxy2/jdk.proxy2.$Proxy47.getListing(Unknown Source)
	at org.apache.hadoop.hdfs.DFSClient.listPaths(DFSClient.java:1733)
	at org.apache.hadoop.hdfs.DFSClient.listPaths(DFSClient.java:1717)
	at org.apache.hadoop.hdfs.DistributedFileSystem.listStatusInternal(DistributedFileSystem.java:1122)
	at org.apache.hadoop.hdfs.DistributedFileSystem$24.doCall(DistributedFileSystem.java:1197)
	at org.apache.hadoop.hdfs.DistributedFileSystem$24.doCall(DistributedFileSystem.java:1194)
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81)
	at org.apache.hadoop.hdfs.DistributedFileSystem.listStatus(DistributedFileSystem.java:1204)
	at org.apache.hadoop.fs.Globber.listStatus(Globber.java:128)
	at org.apache.hadoop.fs.Globber.doGlob(Globber.java:291)
	at org.apache.hadoop.fs.Globber.glob(Globber.java:202)
	at org.apache.hadoop.fs.FileSystem.globStatus(FileSystem.java:2226)
	at org.apache.spark.deploy.SparkHadoopUtil.globPath(SparkHadoopUtil.scala:240)
	at org.apache.spark.sql.execution.datasources.DataSource$.$anonfun$checkAndGlobPathIfNecessary$3(DataSource.scala:816)
	at org.apache.spark.util.ThreadUtils$.$anonfun$parmap$2(ThreadUtils.scala:470)
	at scala.concurrent.Future$.$anonfun$apply$1(Future.scala:691)
	at scala.concurrent.impl.Promise$Transformation.run(Promise.scala:500)
	at java.base/java.util.concurrent.ForkJoinTask$RunnableExecuteAction.exec(ForkJoinTask.java:1395)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)


The three monthly files share the same schema, so one path with a wildcard reads all of them. Together they have 9,554,778 trips and 19 columns. Pickup and dropoff come as `timestamp_ntz`, distance and fares as `double`, and the location IDs as `integer`.

### 0.3 - Quality Checks

Before cleaning anything, I checked the date range, how many trips have zero or negative distance or duration (most knowing as null_count).I also checked how many last more than 4 hours,and the median and 99th percentile of the distance.

In [7]:
# Comment this code for the future 
dur = F.unix_timestamp("tpep_dropoff_datetime") - F.unix_timestamp("tpep_pickup_datetime")

df.select(F.min("tpep_pickup_datetime").alias("pickup_min"),
          F.max("tpep_pickup_datetime").alias("pickup_max"),
          F.avg((F.col("trip_distance") <= 0).cast("int")).alias("dist_zero_ou_neg"),
          F.avg((dur <= 0).cast("int")).alias("dur_zero_ou_neg"),
          F.avg((dur > 4 * 3600).cast("int")).alias("dur_maior_4h"),
          F.expr("percentile_approx(trip_distance, array(0.5, 0.99))").alias("dist_p50_p99")
         ).show(vertical=True, truncate=False)

NameError: name 'df' is not defined

It was found that 2.26% of the trips have zero or negative distance, 0.03% have zero or negative duration and 0.06% last more than 4 hours.  Also the median distance is 1.7 miles and the 99th percentile is 19.87 miles, so most trips are considere short.

### 0.4 Missing Values, Find Outliers Distances and Unknown Zones

This cell measures the share of nulls in the columns that could be features, the largest distance in the data and how many trips start or end in the two unknown zones (which can be called 264 and 265 in the zone lookup table).

In [17]:
# Distancia media - Medicao de numeros nulos 

cols = ["passenger_count", "RatecodeID", "PULocationID", "DOLocationID", "trip_distance"]
df.select([F.avg(F.col(c).isNull().cast("int")).alias(c) for c in cols]).show(vertical=True)

df.select(F.max("trip_distance").alias("dist_max"),
          F.avg(F.col("PULocationID").isin(264, 265).cast("int")).alias("pu_desconhecido"),
          F.avg(F.col("DOLocationID").isin(264, 265).cast("int")).alias("do_desconhecido")
         ).show(vertical=True)

zones = spark.read.option("header", True).csv("hdfs://localhost:9000/ca1/raw/taxi_zone_lookup.csv")
zones.filter(F.col("LocationID").cast("int") >= 264).show(truncate=False)

-RECORD 0------------------------------
 passenger_count | 0.07870010166641235 
 RatecodeID      | 0.07870010166641235 
 PULocationID    | 0.0                 
 DOLocationID    | 0.0                 
 trip_distance   | 0.0                 

-RECORD 0-------------------------------
 dist_max        | 312722.3             
 pu_desconhecido | 0.003847289806210... 
 do_desconhecido | 0.008883408908087661 

+----------+-------+--------------+------------+
|LocationID|Borough|Zone          |service_zone|
+----------+-------+--------------+------------+
|264       |Unknown|N/A           |N/A         |
|265       |N/A    |Outside of NYC|N/A         |
+----------+-------+--------------+------------+



it was found that `passenger_count` and `RatecodeID` are null in 7.87% of the trips, the largest distance is 312,722.3 miles, which is consider a outlier. Also 0.38% of pickups and 0.89% of dropoffs are in unknown zones.

The features chosen for the network are distance, hour of day and day of week from the pickup, and the pickup and dropoff zones. `passenger_count` and `RatecodeID` are not used, so no imputation is needed.

### 0.5 - Cleaning Data

The rules are applied in sequence, on the original values, and each line shows how many trips it has been removed:

* Pickup between 1 January and 31 March 2024
* Distance above 0 and up to 100 miles (the 99th percentile is 19.87 miles, so 100 keeps long airport trips and removes typing errors)
* Duration between 60 seconds and 4 hours
* Speed up to 80 mph, which removes trips that are physically impossible in New York
* Known pickup and dropoff zones.

In [18]:
dur = F.unix_timestamp("tpep_dropoff_datetime") - F.unix_timestamp("tpep_pickup_datetime")
t_ini = F.lit("2024-01-01 00:00:00").cast("timestamp_ntz")
t_fim = F.lit("2024-04-01 00:00:00").cast("timestamp_ntz")

filtros = [
    ("periodo jan-mar 2024", (F.col("tpep_pickup_datetime") >= t_ini) & (F.col("tpep_pickup_datetime") < t_fim)),
    ("distancia (0, 100] mi", (F.col("trip_distance") > 0) & (F.col("trip_distance") <= 100)),
    ("duracao [60 s, 4 h]", (dur >= 60) & (dur <= 4 * 3600)),
    ("velocidade <= 80 mph", F.col("trip_distance") / (dur / 3600) <= 80),
    ("zonas conhecidas", ~F.col("PULocationID").isin(264, 265) & ~F.col("DOLocationID").isin(264, 265)),
]

atual = df
n_ant = df.count()
print(f"{'inicio':<25}{n_ant:>12,}")
for nome, cond in filtros:
    atual = atual.filter(cond)
    n = atual.count()
    print(f"{nome:<25}{n:>12,}   removidas: {n_ant - n:>9,} ({(n_ant - n) / n_ant:.2%})")
    n_ant = n

inicio                      9,554,778
periodo jan-mar 2024        9,554,757   removidas:        21 (0.00%)
distancia (0, 100] mi       9,338,741   removidas:   216,016 (2.26%)
duracao [60 s, 4 h]         9,298,539   removidas:    40,202 (0.43%)
velocidade <= 80 mph        9,298,271   removidas:       268 (0.00%)
zonas conhecidas            9,213,529   removidas:    84,742 (0.91%)


The cleaning process was completed and it left 9,213,529 trips and removes 3.57% of the raw data. 

The filters that remove the most are distance (216,016 trips) and unknown zones (84,742).

### 0.6 - Saving the Clean Data into HDFS. 

In [19]:
dur = F.unix_timestamp("tpep_dropoff_datetime") - F.unix_timestamp("tpep_pickup_datetime")
clean = atual.withColumn("duration_s_true", dur)
clean.write.mode("overwrite").parquet("hdfs://localhost:9000/ca1/clean/yellow_2024q1")

chk = spark.read.parquet("hdfs://localhost:9000/ca1/clean/yellow_2024q1")
print("linhas:", chk.count())
chk.select(F.min("duration_s_true").alias("dur_min"),
           F.max("duration_s_true").alias("dur_max"),
           F.min("tpep_pickup_datetime").alias("pickup_min"),
           F.max("tpep_pickup_datetime").alias("pickup_max")).show(vertical=True)

linhas: 9213529
-RECORD 0-------------------------
 dur_min    | 60                  
 dur_max    | 14344               
 pickup_min | 2024-01-01 00:00:00 
 pickup_max | 2024-03-31 23:59:59 



The clean base has 9,213,529 trips, with durations between 60 and 14,344 seconds and every pickup in the year of 2024. The 3.57% removed are probably data that doesnt express real real trips, thefore, it does not affect future analisys.